### ISSUE TITLE
Default values for nullable enums are not calculated correctly

### ISSUE THREAD

**Issue by mkArtakMSFT** [aspnet/AspNetCore#3165]
_From @JanEggers on April 20, 2018 6:54_

Consider the following code:

```C#
public IActionResult Action(DayOfWeek? day = DayOfWeek.Friday) => null;
```

When the parameter `day` does not get model bound either because no value was specified or if the value on the wire cannot be bound to the enum, executing the action results in an `InvalidCastException`. The issue stems from `ParameterInfo.DefaultValue` returning the numeric value of the default value (i,e, `DayOfWeek.Friday`). We should address this in [`ParameterDefaultValue`](https://github.com/aspnet/DependencyInjection/blob/dev/shared/Microsoft.Extensions.ParameterDefaultValue.Sources/ParameterDefaultValue.cs#L9) to special case default values for nullable enums. Perhaps other nullable types are similarly affected by the behavior of `ParameterInfo.DefaultValue`.

-----------------------------------------
### Original issue

im getting the following exception in production:

System.InvalidCastException: Specified cast is not valid.
at lambda_method(Closure , Object , Object[] )
   at Microsoft.Extensions.Internal.ObjectMethodExecutor.Execute(Object target, Object[] parameters)
 at Microsoft.AspNetCore.Mvc.Internal.ActionMethodExecutor.SyncActionResultExecutor.Execute(ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.<InvokeActionMethodAsync>d__12.MoveNext()

is there any way to figure out what cast is not valid/ what method should have be called/ what is the request that is processed? A little more context would be nice! 

_Copied from original issue: aspnet/Mvc#7692_

**Comment by mkArtakMSFT** [comment]
_From @pranavkm on April 20, 2018 16:48_

It's likely this line: https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/Internal/ActionMethodExecutor.cs#L79. Do you know what action's being executed and possibly the signature of the method? We're likely inferring the return type incorrectly.

**Comment by mkArtakMSFT** [comment]
_From @JanEggers on April 21, 2018 5:59_

> Do you know what action's being executed and possibly the signature of the method?

no i dont know. i see the requestid in the logfiles is there any way too print the info message for that request which prints the uri being requested?

if i enable info logging my logs will explode...

**Comment by mkArtakMSFT** [comment]
_From @JanEggers on April 23, 2018 8:44_

I was able to switch to info logging and got this:

2018-04-23 11:39:41.971 +02:00	[inf]	Executing action method "Devices.Mqtt.Pif.Http.Server.Controllers.PifController.GetRootObjectDefinitions (Devices.Mqtt.Pif.Http.Server)" with arguments (["de", "0"]) - Validation state: Valid	[Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker]
2018-04-23 11:39:41.971 +02:00	[inf]	Executed action "Devices.Mqtt.Pif.Http.Server.Controllers.PifController.GetRootObjectDefinitions (Devices.Mqtt.Pif.Http.Server)" in 1.488ms	[Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker]



```
        [SwaggerOperation( nameof( GetRootObjectDefinitions ) )]
        [Produces( typeof( IEnumerable<ObjectDefinitionViewModel> ) )]
        [HttpGet( "RootObjectDefinitions/{cultureId}" )]
        public IActionResult GetRootObjectDefinitions( 
            string cultureId,
            [FromQuery] ENameDisplayMode? nameDisplayMode = ENameDisplayMode.Name
            )
        {
            return Ok( ...);
        }
```

**Comment by mkArtakMSFT** [comment]
_From @JanEggers on April 23, 2018 9:6_

what is also strange on my machine i get slightly different logs:

2018-04-23 11:03:45.448 +02:00	[inf]	Executing action method "Devices.Mqtt.Pif.Http.Server.Controllers.PifController.GetRootObjectDefinitions (Devices.Mqtt.Pif.Http.Server)" with arguments (["de-CH", "EquipmentIdentificationCode"]) - Validation state: Valid	[Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker]
2018-04-23 11:03:45.448 +02:00	[inf]	Executed action method "Devices.Mqtt.Pif.Http.Server.Controllers.PifController.GetRootObjectDefinitions (Devices.Mqtt.Pif.Http.Server)", **returned result "Microsoft.AspNetCore.Mvc.OkObjectResult"** in 0.0339ms.	[Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker]
2018-04-23 11:03:45.448 +02:00	[inf]	Executing ObjectResult, writing value of type '"System.Collections.Generic.List`1[[Devices.Mqtt.Pif.Models.View.ObjectDefinitionViewModel, Devices.Mqtt.Pif, Version=1.3.0.0, Culture=neutral, PublicKeyToken=null]]"'.	[Microsoft.AspNetCore.Mvc.Infrastructure.ObjectResultExecutor]

**Comment by mkArtakMSFT** [comment]
_From @JanEggers on April 23, 2018 9:31_

I guess it must be related to nullable enum provided as int instead of string

**Comment by mkArtakMSFT** [comment]
_From @JanEggers on April 23, 2018 12:32_

I finally found the issue the enum was provided as "0" (numeric but string) which can not be casted to that enum. it was a browsersession  with an old client running. 

Anyway expected behavior would be that either the cast succeeds or the modelstate should indicate an error but invalidcast exception without context is a little hard to debug.

**Comment by mkArtakMSFT** [comment]
_From @pranavkm on April 23, 2018 17:17_

I'll update the title and we'll triage this as a possible bug.

**Comment by mkArtakMSFT** [comment]
_From @pranavkm on April 24, 2018 1:3_

cc @pakrym for `ParameterDefaultValue` impact. We should try and address the issue there and add a couple of tests to Mvc. @JanEggers for the time being, I'd recommend using a regular \ non-nullable enum in this particular case.

**Comment by kevin-david** [comment]
Hi all, any plan to fix this issue in an upcoming release? I'm currently using ASP.NET Core 1.0 - is it fixed in 2.0 or 2.1? 

I spent a few solid hours debugging this before using the magic Google-fu to figure out where the `InvalidCastException` was coming from and the workaround became obvious - have a dummy `enum` value instead of using a nullable one.

For SEO purposes hopefully including `RouterMiddleware` in my comment will help somebody solve this issue sooner.
